test/libsolidity/ASTJSONTest.h

```diff
@@ -37,6 +38,28 @@ namespace solidity::frontend::test
 class ASTJSONTest: public TestCase
 {
 public:
+	struct TestVariant
+	{
+		TestVariant(std::string _baseName, std::string _variation, CompilerStack::State _stopAfter):
+			variation(_variation),
+			stopAfter(_stopAfter),
+			astFilename(
+				_baseName +
+				(_variation.empty() ? "" : "_") +
+				_variation +
+				".json"
+			)
+		{}
+
+		std::string variation;
+		CompilerStack::State stopAfter;
+		std::string expectation;
```

Comment:
I'd make it optional. The corresponding AST file might not exist.
```suggestion
		std::optional<std::string> expectation;
```

Reply:
I don't think that case exists in practice

Reply:
Then I'd throw an error when the file does not exist (actually, if you use `readFileAsString()` as I suggested you can just drop the `exists()` check and the function will throw).

And then the `empty()` checks (in particular the one I commented on in https://github.com/ethereum/solidity/pull/12200#discussion_r737758783) can be completely removed (replaced with asserts).